### DISCUSSION TITLE
Device Groups do not listen to CCT Color updates

### DISCUSSION THREAD

**Discussion by sillyfrog** [arendst/Tasmota#12012] [Ideas]
### PROBLEM DESCRIPTION

As per a discussion in the #beginer-questions discord channel, it looks like CCT lights do not update correctly when part of the same `DevGroupName`.

Edit/Update: My original setup is 2x CCT lights setup in a `DevGroup`. I can update the brightness and power state OK, and they stay in sync. However if I change the CT, it does _not_ update on the other lights. To debug further, I configured one to be RGBCCT, and found that the CCT updates could at least sync in one direction as follows:

My setup is now as follows:
- Light A is configured correctly as a CCT light, with the `DevGroupName` of `pair`
- Light B is configured as a RGBCCT, with the CCT pins (4/5) connected to the actual CW and WW lights (the RGB pins are assigned to unused pins). This also has a `DevGroupName` of `pair`.
- Both lights are now updated to `9.4.0.3` (but I have seen this issue on different lights since `9.3.1`)

If I adjust the CCT on light A, the corresponding CCT also updates on light B (as expected). The same is also true of brightness and state (on/off). However if I update the CCT on light B, this does _not_ update light A as expected (however brightness and state continues to work).

I started with both lights configured just for CCT, and the color never updated from one to the other (again, other state information did work as expected).

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used: [Arlec Downlight](https://templates.blakadder.com/arlec_ALD092CHA.html) and [BrilliantSmart downlight](https://templates.blakadder.com/brilliantsmart_2069505.html)
- [x] Tasmota binary firmware version number used: v9.3.1, v9.4.0, dev 9.4.0.3.
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: `esptool.py` (erasing before flashing), Tasmotizer (which also uses `esptool.py`) and OTA updates.
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:

**Light A**
```lua
22:06:33.531 CMD: Backlog Template; Module; GPIO 255
22:06:33.555 MQT: stat/tasmota_57/RESULT = {"NAME":"Arlec ALD092CHA","GPIO":[0,0,0,0,416,417,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":18}
22:06:33.760 MQT: stat/tasmota_57/RESULT = {"Module":{"0":"Arlec ALD092CHA"}}
22:06:33.969 MQT: stat/tasmota_57/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"416":"PWM1"},"GPIO5":{"417":"PWM2"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}
```

**Light B**
```lua
22:06:08.738 MQT: stat/tasmota_56/RESULT = {"NAME":"Arlec ALD092CHA","GPIO":[0,0,0,0,419,420,0,0,0,0,416,417,418,0],"FLAG":0,"BASE":18}
22:06:08.938 MQT: stat/tasmota_56/RESULT = {"Module":{"0":"Arlec ALD092CHA"}}
22:06:09.151 MQT: stat/tasmota_56/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"419":"PWM4"},"GPIO5":{"420":"PWM5"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"416":"PWM1"},"GPIO15":{"417":"PWM2"},"GPIO16":{"418":"PWM3"},"GPIO17":{"0":"None"}}
```

- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:
```

- [x] Provide the output of this command: `Status 0`:

**Light A**
```lua
22:08:55.990 MQT: stat/tasmota_57/STATUS = {"Status":{"Module":0,"DeviceName":"Tasmota 57","FriendlyName":["Tasmota"],"Topic":"tasmota_57","ButtonTopic":"0","Power":1,"PowerOnState":1,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":0,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
22:08:55.995 MQT: stat/tasmota_57/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://static.sf/tasmota/dev-9.4.0.3/tasmota.bin","RestartReason":"Software/System restart","Uptime":"0T00:16:09","StartupUTC":"2021-05-04T11:52:46","Sleep":50,"CfgHolder":4617,"BootCount":14,"BCResetTime":"2021-05-04T12:50:05","SaveCount":116,"SaveAddress":"FB000"}}
22:08:56.000 MQT: stat/tasmota_57/STATUS2 = {"StatusFWR":{"Version":"9.4.0.3(tasmota)","BuildDateTime":"2021-05-04T10:33:19","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"341/699"}}
22:08:56.006 MQT: stat/tasmota_57/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["sillyfrog-iot",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008008","2805C8000100060000005A0A000000000000","00000088","00006008","00000000"]}}
22:08:56.015 MQT: stat/tasmota_57/STATUS4 = {"StatusMEM":{"ProgramSize":599,"Free":404,"Heap":22,"ProgramFlashSize":1024,"FlashSize":2048,"FlashChipId":"1540C8","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45","Sensors":"1,2,3,4,5,6"}}
22:08:56.022 MQT: stat/tasmota_57/STATUS5 = {"StatusNET":{"Hostname":"tasmota_57-0032","IPAddress":"10.11.10.57","Gateway":"10.11.255.254","Subnetmask":"255.255.0.0","DNSServer":"10.11.255.254","Mac":"84:CC:A8:B2:C0:20","Webserver":2,"WifiConfig":4,"WifiPower":17.0}}
22:08:56.029 MQT: stat/tasmota_57/STATUS6 = {"StatusMQT":{"MqttHost":"mqtt.sf","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_B2C020","MqttUser":"DVES_USER","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
22:08:56.037 MQT: stat/tasmota_57/STATUS7 = {"StatusTIM":{"UTC":"2021-05-04T12:08:56","Local":"2021-05-04T22:08:56","StartDST":"2021-03-28T02:00:00","EndDST":"2021-10-31T03:00:00","Timezone":"+10:00","Sunrise":"06:14","Sunset":"17:13"}}
22:08:56.044 MQT: stat/tasmota_57/STATUS10 = {"StatusSNS":{"Time":"2021-05-04T22:08:56"}}
22:08:56.050 MQT: stat/tasmota_57/STATUS11 = {"StatusSTS":{"Time":"2021-05-04T22:08:56","Uptime":"0T00:16:10","UptimeSec":970,"Heap":22,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":1,"POWER":"ON","Dimmer":15,"Color":"0422","White":15,"CT":461,"Channel":[2,13],"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"sillyfrog-iot","BSSId":"7E:8A:20:57:2D:0E","Channel":4,"RSSI":96,"Signal":-52,"LinkCount":1,"Downtime":"0T00:00:03"}}}
```

**Light B**
```lua

22:08:03.811 MQT: stat/tasmota_56/STATUS = {"Status":{"Module":0,"DeviceName":"Tasmota 56","FriendlyName":["Tasmota"],"Topic":"tasmota_56","ButtonTopic":"0","Power":1,"PowerOnState":1,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":0,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
22:08:03.818 MQT: stat/tasmota_56/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://static.sf/tasmota/dev-9.4.0.3/tasmota.bin","RestartReason":"Software/System restart","Uptime":"0T00:13:41","StartupUTC":"2021-05-04T11:54:22","Sleep":50,"CfgHolder":4617,"BootCount":16,"BCResetTime":"2021-05-04T13:53:37","SaveCount":152,"SaveAddress":"F8000"}}
22:08:03.823 MQT: stat/tasmota_56/STATUS2 = {"StatusFWR":{"Version":"9.4.0.3(tasmota)","BuildDateTime":"2021-05-04T10:33:19","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"397/699"}}
22:08:03.827 MQT: stat/tasmota_56/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["sillyfrog-iot",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008008","2805C8000100060000005A0A000000000000","00000088","00006008","00000000"]}}
22:08:03.836 MQT: stat/tasmota_56/STATUS4 = {"StatusMEM":{"ProgramSize":599,"Free":404,"Heap":22,"ProgramFlashSize":1024,"FlashSize":2048,"FlashChipId":"1540C8","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45","Sensors":"1,2,3,4,5,6"}}
22:08:03.843 MQT: stat/tasmota_56/STATUS5 = {"StatusNET":{"Hostname":"tasmota_56-0218","IPAddress":"10.11.10.56","Gateway":"10.11.255.254","Subnetmask":"255.255.0.0","DNSServer":"10.11.255.254","Mac":"84:CC:A8:B1:E0:DA","Webserver":2,"WifiConfig":4,"WifiPower":17.0}}
22:08:03.849 MQT: stat/tasmota_56/STATUS6 = {"StatusMQT":{"MqttHost":"mqtt.sf","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_B1E0DA","MqttUser":"DVES_USER","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
22:08:03.857 MQT: stat/tasmota_56/STATUS7 = {"StatusTIM":{"UTC":"2021-05-04T12:08:03","Local":"2021-05-04T22:08:03","StartDST":"2021-03-28T02:00:00","EndDST":"2021-10-31T03:00:00","Timezone":"+10:00","Sunrise":"06:14","Sunset":"17:13"}}
22:08:03.863 MQT: stat/tasmota_56/STATUS10 = {"StatusSNS":{"Time":"2021-05-04T22:08:03"}}
22:08:03.871 MQT: stat/tasmota_56/STATUS11 = {"StatusSTS":{"Time":"2021-05-04T22:08:03","Uptime":"0T00:13:41","UptimeSec":821,"Heap":22,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":1,"POWER":"ON","Dimmer":15,"Color":"0000000422","HSBColor":"0,0,0","White":15,"CT":461,"Channel":[0,0,0,2,13],"Scheme":0,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"sillyfrog-iot","BSSId":"7E:8A:20:57:2D:0E","Channel":4,"RSSI":94,"Signal":-53,"LinkCount":1,"Downtime":"0T00:00:03"}}}
```

- [x] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:

**Light B**
```lua
22:12:25.358 CMD: Group 0, Index 1, Command "WEBLOG", Data "4"
22:12:25.364 MQT: stat/tasmota_57/RESULT = {"WebLog":4}
22:12:28.409 DGR: Sending pair message to 10.11.10.56: seq=15, flags=8
22:12:28.414 DGR: Received pair message from 10.11.10.56: seq=15, flags=0, 224=0,0,0,16,239,14*
22:12:28.416 DGR: Sending pair message to 10.11.10.56: seq=15, flags=8
22:12:28.419 DGR: Received pair message from 10.11.10.56: seq=15, flags=0 (old)
22:12:30.503 WIF: Checking connection...
```

### TO REPRODUCE
_Steps to reproduce the behavior:_

1. Configure 2 CCT lights, with just 2 PWM channels, and configure them to be part of the same `DevGroupName`.
2. Check that brightness and power state (on/off) is synchronising correctly.
3. Adjust the color temperature on one light, it will _not_ update on the other as expected.

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

In step 3 above, I would expect the color temperature to also update.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

N/A

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

I have seen this across a few different lights now (including ones that have a template base `48` - one PWM channel is brightness, and the other is the "color temperature"), but this is the first time I have had the chance to test in detail. The lights I'm testing on are one channel for cool white and the other for warm white (the more normal way I think).

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by sillyfrog** [comment]
As this has been moved to a feature request; I'm keen to understand how this is different from the current color synchronizing (ie: what's different, and how should I craft this as a feature request)? Obviously currently RGB synchronises, as does RGBCCT. Should we update the documentation at https://tasmota.github.io/docs/Device-Groups/ to clarify what the goals are of the current design (or at least what is not included)?
Thanks!

**Reply by pcdiem** [reply]
The issue is that the light module function that is used to change the channels remaps CCT lights from channels 1 and 2 instead of using channels 4 and 5 as they arrive from the network. I'll submit a PR to map channels 4 and 5 to channels 1 and 2 before calling the function.

**Reply by sillyfrog** [reply]
Thanks! I was about to write that I think it maybe something like that :) 
Testing just now I noticed that changing the Red and Green on Light B updates the CCT on Light A, implying some sort of issue along those lines.
Let me know if I can help/test or anything else.
Thanks again!

**Reply by pcdiem** [reply]
What would we expect to happen when a CCT light is in a device group with an RGBCCT light and the RGBCCT light is set to red? Would we expect the CCT light to turn off since there are no white channels on?

**Reply by sillyfrog** [reply]
Personally, I would say it should follow the CCT channels, so if they are off on the RGBCCT, they also turn off on the CCT lights. I would have thought it wouldn't be super common to have RGBCCT and CCT all mixed and grouped, so hopefully not a big issue.
Maybe it needs to be it's own bitmask so the user can select? eg: have one bit for RGB, and one for CCT? But that's (potentially) a breaking change.

**Reply by sfromis** [reply]
You could also wonder about how the CT propagation would work if the DG includes devices with ChannelRemap (SetOption37) independent handling of RGB and white channels, and also virtual CT channels (mixing RGB and white to aim for CT-like variation of color), but this may be getting a bit far out in the corners.

**Reply by sillyfrog** [reply]
Thinking about this more today... Given the current CT lights automatically map their 1/2 channels to 4/5, then I think doing the reverse on the way back in would be most logical. ie: if you have a CT light, and a `DevGroup` update comes in, then channels 4/5 should be mapped back to 1/2 (like you originally said @pcdiem ).
Please let me know if there is anything I can do to help.
Thanks!

**Reply by pcdiem** [reply]
Currently the code just sends and received the light channels in the order they are stored on the local device. This, of course, doesn't map correctly when using a mix of W, CW, RGB, RGBW and RGBCW lights. I guess the correct thing to do is to make the device group light channels always be R, G, B, C, W.

Under the proposal, a W only light will send 0, 0, 0, W, 0 for the light channels; a CW light will send/receive 0, 0, 0, C, W; an RGB light will send/receive R, G, B, 0, 0; an RGBW light will send/receive R, G, B, W, 0; an RGBCW light will send/receive R, G, B, C, W. 

Note that for W and RGBW lights, we will send the white channel as cold white since we don't know what temperature the white actually is.

For W and CW lights, when a light channel update is received that has the C and W channels as 0, we could set the color temperature based on the R, G, B channels if the mix of those channels is close to white. This seems a bit complex though.

Assuming DevGroupShare is set to send and received light channels on all lights in the group, this new logic will result in all varieties of lights in the group keeping cold white, warm white and colors in sync with the following things to be noted:
- When an RGB, RGBW or RGBCW light is set to a color, all W and CW lights in the group will be turned off (actually just dimmed to 0).
- When a W light is turned on or a CW or RGBCW light is set to a color temperature, all RGB lights in the group will be turned off (actually just dimmed to 0).
- When a W light is turned on or an RGBW light is set to white, all CW and RGBCW lights in the group will be changed to cold white.

Does this seem reasonable?

**Reply by sillyfrog** [reply]
Before going down this path too far - to clarify as rereading the original issue I realize I missed something super important (the reason this started this issue - I came in after the discussion on Discord and it was obvious in my head): What I actually want is to have all CCT lights stay in sync. At the moment this does not work.
My original setup is 2x CCT lights, and when adjusting the CT they did not update (only brightness and power state). The reason I created the dummy RGBCCT is to test things and find something that sort of worked.
So maybe we should go back to the start and figure out why a pure CCT setup is not working if there is no channel mapping going on by design as the source of the issue maybe a bug else where.

I'm assuming there is channel mapping going on because in the CCT / RGBCCT setup, the CCT light can update the CCT values on the RGBCCT light.

Hopefully that makes more sense? (I'll edit the original description now as well)

**Reply by sillyfrog** [reply]
Further to the above, this is the weblogs from the lights when changing Light A from warm white to cool white, and back to warm white when both are setup correctly as a CCT light. (Light B does _not_ get the CCT update)

**Light A**
```lua
07:05:41.474 SRC: WebGui from 10.1.10.2
07:05:41.477 CMD: Group 0, Index 1, Command "CT", Data "153"
07:05:41.484 MQT: stat/tasmota_57/RESULT = {"POWER":"ON","Dimmer":50,"Color":"8000","White":50,"CT":153,"Channel":[50,0]}
07:05:41.508 DGR: Sending pair message to network: seq=12, flags=0, 224=0,0,0,255,0,12
07:05:41.682 CFG: Saved to flash at F6, Count 233, Bytes 4096
07:05:41.718 DGR: Sending pair message to 10.11.10.56: seq=12, flags=0, last ack=11
07:05:41.748 DGR: Received pair message from 10.11.10.56: seq=12, flags=8
07:05:41.788 DGR: Received pair message from 10.11.10.56: seq=12, flags=8
07:05:43.625 SRC: WebGui from 10.1.10.2
07:05:43.629 CMD: Group 0, Index 1, Command "CT", Data "500"
07:05:43.636 MQT: stat/tasmota_57/RESULT = {"POWER":"ON","Dimmer":50,"Color":"0080","White":50,"CT":500,"Channel":[0,50]}
07:05:43.656 DGR: Sending pair message to network: seq=13, flags=0, 224=0,0,0,0,255,13
07:05:43.731 CFG: Saved to flash at F5, Count 234, Bytes 4096
07:05:43.737 DGR: Received pair message from 10.11.10.56: seq=13, flags=8
```


**Light B**
```lua
07:05:41.961 DGR: Sending pair message to 10.11.10.57: seq=12, flags=8
07:05:41.965 DGR: Received pair message from 10.11.10.57: seq=12, flags=0, 224=0,0,0,255,0,12*
07:05:41.001 DGR: Sending pair message to 10.11.10.57: seq=12, flags=8
07:05:41.004 DGR: Received pair message from 10.11.10.57: seq=12, flags=0 (old)
07:05:43.903 DGR: Sending pair message to 10.11.10.57: seq=13, flags=8
07:05:43.907 DGR: Received pair message from 10.11.10.57: seq=13, flags=0, 224=0,0,0,0,255,13*
```

In those updates (best I can tell), it looks like channels 4 and 5 are getting updated (not 1 and 2) based on the portions such as: `224=0,0,0,255,0,12*`. (ie: the `255,0` before the sequence number).

Where as if I have one setup as RGBCCT, and I adjust to (mostly) green, then back to red, it sends the following:
```lua
07:13:57.817 SRC: WebGui from 10.1.10.2
07:13:57.820 CMD: Group 0, Index 1, Command "HSBCOLOR", Data "112"
07:13:57.828 MQT: stat/tasmota_56/RESULT = {"POWER":"ON","Dimmer":67,"Color":"17AB000000","HSBColor":"112,100,67","White":0,"CT":500,"Channel":[9,67,0,0,0]}
07:13:57.850 DGR: Sending pair message to network: seq=12, flags=0, 224=34,255,0,0,0,10
07:13:58.066 DGR: Sending pair message to 10.11.10.57: seq=12, flags=0, last ack=11
07:13:58.083 DGR: Received pair message from 10.11.10.57: seq=12, flags=8
07:13:58.124 DGR: Received pair message from 10.11.10.57: seq=12, flags=8
07:13:58.572 CFG: Saved to flash at FA, Count 294, Bytes 4096
07:13:59.756 WIF: Checking connection...
07:14:03.975 SRC: WebGui from 10.1.10.2
07:14:03.979 CMD: Group 0, Index 1, Command "HSBCOLOR", Data "1"
07:14:03.987 MQT: stat/tasmota_56/RESULT = {"POWER":"ON","Dimmer":67,"Color":"AB03000000","HSBColor":"1,100,67","White":0,"CT":500,"Channel":[67,1,0,0,0]}
07:14:04.020 DGR: Sending pair message to network: seq=13, flags=0, 224=255,4,0,0,0,11
07:14:04.225 DGR: Sending pair message to 10.11.10.57: seq=13, flags=0, last ack=12
07:14:04.265 DGR: Received pair message from 10.11.10.57: seq=13, flags=8
07:14:04.285 DGR: Received pair message from 10.11.10.57: seq=13, flags=8
07:14:04.565 CFG: Saved to flash at F9, Count 295, Bytes 4096
```

This _does_ adjust the light on the CCT light correctly, and appears to be updating channel 1 and 2 (specifically the `224=34,255,0,0,0,10` and `224=255,4,0,0,0,11`.

Does that help clarify things further? I'm _really_ sorry for not including all of this in the original issue and leading you down the wrong path.

In short, after our discussions, I think the problem is that something is mapping the CCT lights to channels 4 and 5 when sending to the `DevGroup` in an unexpected way. So my thinking is we should either have it map it back to 1 and 2 (like @pcdiem has proposed), or figure out why it's mapping to 4 and 5, and fix that. Does that make sense?

**Comment by sillyfrog** [comment]
I have been looking around the source code further to try and figure this issue out and what the source of the issue is.

I have been looking in `xdrv_04_light.ino`

I think the issue is that in `LightSendDeviceGroupStatus()`, on line 2192 it uses the Raw channels:
```c
    light_state.getChannelsRaw(channels);
```

Which is simply:
```c
    void getChannelsRaw(uint8_t *channels) {
      channels[0] = _r;
      channels[1] = _g;
      channels[2] = _b;
      channels[3] = _wc;
      channels[4] = _ww;
    }
```

Where as, in `LightHandleDevGroupItem` line 2261, it uses:
```c
          light_controller.changeChannels(Light.channels_fifo);
```

So the first call to get “channels raw” it is not doing any channel mapping, but when it comes time to set it, in `changeChannels`, we have (line 913): 
```c
    } else if (LST_COLDWARM == Light.subtype) {
      // remap channels 0-1 to 3-4 if cold/warm
      uint8_t remapped_channels[5] = {0,0,0,channels[0],channels[1]};
      _state->setChannels(remapped_channels);
    } else {
```
So it _is_ mapping channels.

In my testing, if I remove the above `else if` statement, things work for Dev Group syncing, but this will have other knock on effects.

Something else that was weird to me, because of the mapping, it should be getting all `0` for each channel, but in `setCW`, there is a special case for `if (0 == max) ` - which will always be true - so the existing CW/WW values don't get changed (which is what I would have expected at first glance).

This finally lead me to the following diff for something that appears to work (but I don’t know the code well, and unsure if there will be any other unexpected side effects, please also ignore the extra logs I have for my debugging - this is just an example for reference):
```diff
diff --git a/tasmota/xdrv_04_light.ino b/tasmota/xdrv_04_light.ino
index f41ace288..c8282f2d1 100644
--- a/tasmota/xdrv_04_light.ino
+++ b/tasmota/xdrv_04_light.ino
@@ -907,12 +907,17 @@ public:
   // set all 5 channels at once.
   // Channels are: R G B CW WW
   // Brightness is automatically recalculated to adjust channels to the desired values
-  void changeChannels(uint8_t *channels) {
+  // if channelsmapped is true, CCT/RGB mapping has already been done
+  void changeChannels(uint8_t *channels, bool channelsmapped=false) {
+    AddLog(LOG_LEVEL_INFO, "ChangeChannels: Light.pwm_multi_channels: %d, Light.subtype: %d", Light.pwm_multi_channels, Light.subtype);
+    AddLog(LOG_LEVEL_INFO, "ChangeChannels: Channels: 0: %d, 1: %d, 2: %d, 3: %d, 4: %d", channels[0], channels[1], channels[2], channels[3], channels[4]);
     if (Light.pwm_multi_channels) {
       _state->setChannelsRaw(channels);
-    } else if (LST_COLDWARM == Light.subtype) {
+    } else if (LST_COLDWARM == Light.subtype && ! channelsmapped) {
+      AddLog(LOG_LEVEL_INFO, "ChangeChannels: Entered LST_COLDWARM");
       // remap channels 0-1 to 3-4 if cold/warm
       uint8_t remapped_channels[5] = {0,0,0,channels[0],channels[1]};
+    AddLog(LOG_LEVEL_INFO, "ChangeChannels: REMAPPED Channels: 0: %d, 1: %d, 2: %d, 3: %d, 4: %d", remapped_channels[0], remapped_channels[1], remapped_channels[2], remapped_channels[3], remapped_channels[4]);
       _state->setChannels(remapped_channels);
     } else {
       _state->setChannels(channels);
@@ -2270,7 +2275,7 @@ void LightHandleDevGroupItem(void)
         }
         else
 #endif  // USE_DGR_LIGHT_SEQUENCE
-          light_controller.changeChannels((uint8_t *)XdrvMailbox.data);
+          light_controller.changeChannels((uint8_t *)XdrvMailbox.data, true);
         light_controller.changeBri(bri);
       }
       send_state = true;
```

This provides a new optional argument to prevent the mapping of channels in `changeChannels`.

What do you think? Should I make a proper PR and submit? Are there any other issues with this? Is there a better way?

Thanks!

**Comment by pcdiem** [comment]
This should work to properly handle keep CW lights in sync. It will not keep RGBCW in sync with CW lights though. Maybe it's simple enough to just note in the docs that syncing lights with different PWM counts is not supported.

In the DGR_ITEM_LIGHT_CHANNELS case block there is another light_controller.changeChannels call just above the one you changed that needs to have the true argument as well:

        if (Light.sequence_offset) {
          light_controller.changeChannels(Light.channels_fifo, true);

**Reply by sfromis** [reply]
While my expectation would indeed be that only lights of the same type (multiple points for the same lighting task) are put in a group..... If I try to speculate about a somewhat plausible setup for mixing, it could be using CCT lights for normal "white" lighting purposes, but having some of the bulbs including RGB for "special purposes", not used at the same time as whites. 

However, I'd not at all be sad about not being able to have this in a combined group, with the complexities arising from how group commands are to apply to different lamps. The mixed case could probably then be handled by a rule doing one-way propagation from the white-only group to the mixed color group, as a perfectly satisfactory solution.

**Reply by sillyfrog** [reply]
I think this will keep everything in sync :)
When a change is broadcast, it uses `getChannelsRaw`, which always allocates channels 1, 2, 3, 4, 5 to R, G, B, CW, WW respectively, and only on CCT lights will it (now) then map CW and WW back to channels 1, 2 on the way in (while continuing to treat them as WW and CW). This means that all broadcasts will always have a known state, irrespective of the type of light (RGB, RGBCCT, CCT etc). I'll do a proper build and get it on my lights (my testing has just been on a Wemos) and confirm what happens.

**Reply by sillyfrog** [reply]
OK, so in my testing, things appear to work as expected (but my understanding of how RGBCCT lights work was wrong). It works in so much as the CCT channels and broadcast correctly, and update on all lights (irrespective of type) as expected. However when updating an RGBCCT light, when selecting a color, the CCT channels are turned off, and only the RGB ones are set (which is what I think you were saying earlier, but I mis-understood how they work). However, because both of the CCT channels are zero, in the `setCW` function we have:
```
    void setCW(uint8_t c, uint8_t w, bool free_range = false) {
      uint16_t max = (w > c) ? w : c;   // 0..255
      uint16_t sum = c + w;
      if (sum <= 257) { free_range = false; }    // if we don't allow free range or if sum is below 255 (with tolerance of 2)

      if (0 == max) {
    ...
```
That `max` means the color is not actually set, so the CCT light remains where it was (ie: what ever CCT color was last sent).

When going from RGB to CCT, the same happens. When going from CCT to RGB, the RGB just ignores the color updates.

But that said, I think this is the best anyone can expect with a mixture of RGB, RGBCCT and CCT lights, and most importantly it all works if you have all lights of the same type.

I'll test this on an ESP32 and submit a PR shortly.

**Reply by sillyfrog** [reply]
That PR is submitted #12044 . If someone is able to please [approve the workflow](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks) to start, that would be great (in case I missed anything, I'll fix ASAP). Thanks.

**Reply by pcdiem** [reply]
Thanks for you work on this.

**Comment by Tediore** [comment]
Found this discussion while I was researching before potentially opening an issue. I've noticed a secondary problem: I have four RGB-CCT lamps in a device group on Tasmota 9.5, and all of them have `ctrange` set to 167,333 to reflect the lamps' actual CCT range. Sending a `ct x` command results in the wrong `ct` value being sent to the other lamps (and the wrong perceived color temperature as a result). For example, when issuing `ct 250` to one lamp, I see `13:11:38.126 MQT: stat/office_nw/RESULT = {"CT":327}` on all of the other lamps which now show the wrong color temp compared to the lamp I issued the command on. It appears the `ct` value sent from the first lamp is scaled to the default range (153,500) on the other lamps in the group for some reason.

I'd be happy to provide other information or open an issue if requested.

**Reply by sillyfrog** [reply]
@Tediore Was this an issue prior to 9.5? Or are you just getting started? (I did the PR to sort out the CCT mapping issue, and it may have affected this as I did not test this specific setup).

**Reply by Tediore** [reply]
Unfortunately I'm not sure; I just started using this particular setup on 9.5. I'm supposed to have more of these bulbs delivered tomorrow--is there a tasmota version I can flash to them instead of 9.5 that would be helpful for testing? Perhaps right after `ctrange` was added?
